pkg/balancer/least_connections.go

```diff
@@ -0,0 +1,129 @@
+package balancer
+
+import (
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/square/p2/pkg/util"
+)
+
+type lcBackend struct {
+	conns int
+
+	// disabling allows us to gracefully release connections without assigning
+	// new connections to the backend.
+	disabled bool
+}
+
+type LeastConnections struct {
+	backends map[string]*lcBackend
+	mapMux   sync.Mutex
+}
+
+func NewLeastConnectionsStrategy(initialAddresses []string) (*LeastConnections, error) {
```

Comment:
I think you can get rid of `initialAddresses`. It's always given a zero value in non-test code, and even for test code, you could get the same behavior by creating a test-only wrapper function that calls `AddAddress()` in a loop.  There's nothing special about this constructor.